File: app/src/main/java/com/nexosolar/android/ui/invoices/FilterFragment.java
package com.nexosolar.android.ui.invoices;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import com.google.android.material.datepicker.CalendarConstraints;
import com.google.android.material.datepicker.MaterialDatePicker;
import com.nexosolar.android.R;
import com.nexosolar.android.core.DateUtils;
import com.nexosolar.android.databinding.FragmentFilterBinding;
import com.nexosolar.android.domain.models.InvoiceFilters;
import com.nexosolar.android.domain.models.InvoiceState;
// AJUSTA ESTE IMPORT según dónde hayas puesto el archivo RangeValidator.java
import com.nexosolar.android.ui.common.RangeValidator;

import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneOffset;
import java.util.ArrayList;
import java.util.List;

/**
 * FilterFragment
 *
 * Fragmento modal que permite aplicar filtros avanzados sobre las facturas.
 * Ofrece filtrado por:
 * - Rango de fechas (inicio y fin)
 * - Rango de importes (slider)
 * - Estados de factura (checkboxes múltiples)
 *
 * Comportamiento de fechas:
 * - UI: Muestra "día/mes/año" cuando no hay fecha explícitamente seleccionada.
 * - Lógica interna: Usa las fechas extremas reales (más antigua/más nueva)
 *   para evitar filtros vacíos innecesarios.
 *
 * Sigue arquitectura MVVM: solo maneja UI y observa el ViewModel compartido.
 */
public class FilterFragment extends Fragment {

    private static final String DATE_PICKER_TAG = "DATE_PICKER";

    // ===== Variables de instancia =====
    private FragmentFilterBinding binding;
    private InvoiceViewModel viewModel;

    // ===== Ciclo de vida =====
    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        binding = FragmentFilterBinding.inflate(inflater, container, false);
        viewModel = new ViewModelProvider(requireActivity()).get(InvoiceViewModel.class);

        setupObservers();
        setupListeners();

        return binding.getRoot();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    // ===== Configuración de observadores =====

    private void setupObservers() {
        viewModel.getFiltrosActuales().observe(getViewLifecycleOwner(), this::actualizarUI);
        viewModel.getErrorValidacion().observe(getViewLifecycleOwner(), error -> {
            if (error != null) {
                Toast.makeText(requireContext(), error, Toast.LENGTH_SHORT).show();
            }
        });
    }

    // ===== Configuración de listeners =====

    @SuppressLint("DefaultLocale")
    private void setupListeners() {
        binding.btnSelectDate.setOnClickListener(v -> abrirDatePicker(true));
        binding.btnSelectDateUntil.setOnClickListener(v -> abrirDatePicker(false));

        binding.rangeSlider.addOnChangeListener((slider, value, fromUser) -> {
            List<Float> values = slider.getValues();
            if (values.size() == 2) {
                float min = values.get(0);
                float max = values.get(1);
                binding.tvMinValue.setText(String.format("%.0f €", min));

                float maxImporte = viewModel.getMaxImporte();
                if (max >= maxImporte - 0.001f) {
                    binding.tvMaxValue.setText(String.format("%.2f €", max));
                } else {
                    binding.tvMaxValue.setText(String.format("%.0f €", max));
                }
            }
        });

        binding.checkPagadas.setOnCheckedChangeListener((v, isChecked) ->
                actualizarEstadoCheckbox(InvoiceState.PAID.serverValue, isChecked));
        binding.checkPendientesPago.setOnCheckedChangeListener((v, isChecked) ->
                actualizarEstadoCheckbox(InvoiceState.PENDING.serverValue, isChecked));
        binding.checkAnuladas.setOnCheckedChangeListener((v, isChecked) ->
                actualizarEstadoCheckbox(InvoiceState.CANCELLED.serverValue, isChecked));
        binding.checkCuotaFija.setOnCheckedChangeListener((v, isChecked) ->
                actualizarEstadoCheckbox(InvoiceState.FIXED_FEE.serverValue, isChecked));
        binding.checkPlanPago.setOnCheckedChangeListener((v, isChecked) ->
                actualizarEstadoCheckbox(InvoiceState.PAYMENT_PLAN.serverValue, isChecked));

        binding.btnAplicar.setOnClickListener(v -> aplicarFiltros());
        binding.btnBorrar.setOnClickListener(v -> viewModel.resetearFiltros());
        binding.btnCerrar.setOnClickListener(v -> cerrarFragmento());
    }

    // ===== Actualización de UI =====

    @SuppressLint("DefaultLocale")
    private void actualizarUI(InvoiceFilters filtros) {
        if (filtros == null) return;

        // Fechas
        if (filtros.startDate != null) {
            binding.btnSelectDate.setText(DateUtils.formatDateShort(filtros.startDate));
        } else {
            binding.btnSelectDate.setText(R.string.dia_mes_ano);
        }

        if (filtros.endDate != null) {
            binding.btnSelectDateUntil.setText(DateUtils.formatDateShort(filtros.endDate));
        } else {
            binding.btnSelectDateUntil.setText(R.string.dia_mes_ano);
        }

        // Slider
        float maxImporte = viewModel.getMaxImporte();
        if (maxImporte > 0) {
            binding.rangeSlider.setValueFrom(0f);
            binding.rangeSlider.setValueTo(maxImporte);

            float minVal = Math.max(0f, filtros.minAmount.floatValue());
            float maxVal = Math.min(maxImporte, filtros.maxAmount.floatValue());
            if (minVal > maxVal) minVal = maxVal;

            binding.rangeSlider.setValues(minVal, maxVal);
            binding.tvMaxImporte.setText(String.format("%.2f €", maxImporte));
            binding.tvMinValue.setText(String.format("%.0f €", minVal));
            binding.tvMaxValue.setText(String.format("%.0f €", maxVal));
        }

        // Checkboxes
        List<String> estados = filtros.filteredStates;
        if (estados == null) estados = new ArrayList<>();
        binding.checkPagadas.setChecked(estados.contains(InvoiceState.PAID.serverValue));
        binding.checkPendientesPago.setChecked(estados.contains(InvoiceState.PENDING.serverValue));
        binding.checkAnuladas.setChecked(estados.contains(InvoiceState.CANCELLED.serverValue));
        binding.checkCuotaFija.setChecked(estados.contains(InvoiceState.FIXED_FEE.serverValue));
        binding.checkPlanPago.setChecked(estados.contains(InvoiceState.PAYMENT_PLAN.serverValue));
    }

    // ===== Gestión de filtros =====

    private void aplicarFiltros() {
        InvoiceFilters nuevosFiltros = new InvoiceFilters();

        // 1. Estados
        List<String> estados = new ArrayList<>();
        if (binding.checkPagadas.isChecked()) estados.add(InvoiceState.PAID.serverValue);
        if (binding.checkPendientesPago.isChecked()) estados.add(InvoiceState.PENDING.serverValue);
        if (binding.checkAnuladas.isChecked()) estados.add(InvoiceState.CANCELLED.serverValue);
        if (binding.checkCuotaFija.isChecked()) estados.add(InvoiceState.FIXED_FEE.serverValue);
        if (binding.checkPlanPago.isChecked()) estados.add(InvoiceState.PAYMENT_PLAN.serverValue);
        nuevosFiltros.setFilteredStates(estados);

        // 2. Importes
        List<Float> valores = binding.rangeSlider.getValues();
        nuevosFiltros.minAmount = (double) valores.get(0);
        nuevosFiltros.maxAmount = (double) valores.get(1);

        // 3. Fechas
        // IMPORTANTE: Aquí recuperamos las fechas que el usuario HA ELEGIDO explícitamente.
        // Si no ha tocado el calendario, seguirán siendo null (lo cual es correcto).
        InvoiceFilters current = viewModel.getFiltrosActuales().getValue();
        LocalDate start = (current != null) ? current.startDate : null;
        LocalDate end = (current != null) ? current.endDate : null;

        // Solo corregimos si AMBAS existen y están cruzadas
        if (start != null && end != null && start.isAfter(end)) {
            LocalDate temp = start;
            start = end;
            end = temp;
        }

        nuevosFiltros.startDate = start;
        nuevosFiltros.endDate = end;

        viewModel.actualizarFiltros(nuevosFiltros);
        cerrarFragmento();
    }


    private void actualizarEstadoCheckbox(String estado, boolean isChecked) {
        InvoiceFilters filtros = viewModel.getFiltrosActuales().getValue();
        if (filtros == null) return;

        List<String> estadosActuales = filtros.filteredStates;
        if (estadosActuales == null) estadosActuales = new ArrayList<>();
        else estadosActuales = new ArrayList<>(estadosActuales);

        if (isChecked) {
            if (!estadosActuales.contains(estado)) estadosActuales.add(estado);
        } else {
            estadosActuales.remove(estado);
        }
        filtros.setFilteredStates(estadosActuales);

        // Preservar valores del slider
        List<Float> currentSliderValues = binding.rangeSlider.getValues();
        if (currentSliderValues.size() >= 2) {
            filtros.minAmount = (double) currentSliderValues.get(0);
            filtros.maxAmount = (double) currentSliderValues.get(1);
        }
        viewModel.actualizarEstadoFiltros(filtros);
    }

    // ===== DatePicker =====

    private void abrirDatePicker(boolean esInicio) {
        InvoiceFilters filtrosActuales = viewModel.getFiltrosActuales().getValue();
        if (filtrosActuales == null) return;

        // 1. Límites Globales (el universo de fechas posibles según los datos)
        LocalDate globalMin = viewModel.getOldestDate();
        LocalDate globalMax = viewModel.getNewestDate();
        long today = MaterialDatePicker.todayInUtcMilliseconds();

        long globalMinMillis = (globalMin != null) ? globalMin.atStartOfDay(ZoneOffset.UTC).toInstant().toEpochMilli() : today;
        long globalMaxMillis = (globalMax != null) ? globalMax.atStartOfDay(ZoneOffset.UTC).toInstant().toEpochMilli() : today;

        // 2. Límites Restringidos (según la selección actual del OTRO campo)
        long constraintMin;
        long constraintMax;

        if (esInicio) {
            // Si elijo inicio, no puedo ir más allá del final seleccionado (si lo hay)
            constraintMin = globalMinMillis;
            if (filtrosActuales.endDate != null) {
                constraintMax = filtrosActuales.endDate.atStartOfDay(ZoneOffset.UTC).toInstant().toEpochMilli();
            } else {
                constraintMax = globalMaxMillis;
            }
        } else {
            // Si elijo fin, no puedo ir antes del inicio seleccionado (si lo hay)
            if (filtrosActuales.startDate != null) {
                constraintMin = filtrosActuales.startDate.atStartOfDay(ZoneOffset.UTC).toInstant().toEpochMilli();
            } else {
                constraintMin = globalMinMillis;
            }
            constraintMax = globalMaxMillis;
        }

        // Defensa contra inconsistencias (min > max)
        if (constraintMin > constraintMax) {
            constraintMin = globalMinMillis;
            constraintMax = globalMaxMillis;
        }

        // 3. CONSTRUCCIÓN DEL VALIDADOR PROPIO
        // Aquí usamos tu RangeValidator que compara LocalDate vs LocalDate
        CalendarConstraints.Builder constraintsBuilder = new CalendarConstraints.Builder()
                .setStart(globalMinMillis) // Esto permite hacer scroll visual hasta el mínimo global
                .setEnd(globalMaxMillis)   // Esto permite hacer scroll visual hasta el máximo global
                .setValidator(new RangeValidator(constraintMin, constraintMax)); // <--- VALIDACIÓN REAL

        // 4. Selección inicial del calendario
        LocalDate fechaActual = esInicio ? filtrosActuales.startDate : filtrosActuales.endDate;
        long selection;
        if (fechaActual != null) {
            selection = fechaActual.atStartOfDay(ZoneOffset.UTC).toInstant().toEpochMilli();
        } else {
            // Si no hay selección, sugerimos una fecha inteligente
            LocalDate smartDate = esInicio ? globalMin : globalMax;
            if (smartDate == null) smartDate = LocalDate.now();
            selection = smartDate.atStartOfDay(ZoneOffset.UTC).toInstant().toEpochMilli();
        }

        // Aseguramos que la selección inicial caiga dentro del rango permitido
        if (selection < constraintMin) selection = constraintMin;
        if (selection > constraintMax) selection = constraintMax;

        constraintsBuilder.setOpenAt(selection);

        MaterialDatePicker<Long> datePicker = MaterialDatePicker.Builder.datePicker()
                .setTitleText(esInicio ? "Seleccionar Inicio" : "Seleccionar Fin")
                .setSelection(selection)
                .setCalendarConstraints(constraintsBuilder.build())
                .setTheme(com.google.android.material.R.style.ThemeOverlay_MaterialComponents_MaterialCalendar)
                .build();

        datePicker.addOnPositiveButtonClickListener(selectedMillis -> {
            LocalDate nuevaFecha = Instant.ofEpochMilli(selectedMillis).atZone(ZoneOffset.UTC).toLocalDate();
            InvoiceFilters filtros = viewModel.getFiltrosActuales().getValue();

            if (filtros != null) {
                if (esInicio) {
                    filtros.startDate = nuevaFecha;
                    // Si la nueva fecha de inicio supera al fin actual, empujamos el fin
                    if (filtros.endDate != null && nuevaFecha.isAfter(filtros.endDate)) {
                        filtros.endDate = nuevaFecha;
                    }
                } else {
                    filtros.endDate = nuevaFecha;
                    // Si la nueva fecha de fin es anterior al inicio actual, empujamos el inicio
                    if (filtros.startDate != null && filtros.startDate.isAfter(nuevaFecha)) {
                        filtros.startDate = nuevaFecha;
                    }
                }
                viewModel.actualizarEstadoFiltros(filtros);
            }
        });

        datePicker.show(getParentFragmentManager(), DATE_PICKER_TAG);
    }

    private void cerrarFragmento() {
        if (getActivity() instanceof InvoiceListActivity) {
            ((InvoiceListActivity) getActivity()).restoreMainView();
        }
        getParentFragmentManager().popBackStack();
    }
}

File: app/src/main/java/com/nexosolar/android/ui/invoices/FilterFragment.kt
package com.nexosolar.android.ui.invoices

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.google.android.material.datepicker.CalendarConstraints
import com.google.android.material.datepicker.MaterialDatePicker
import com.nexosolar.android.R
import com.nexosolar.android.core.DateUtils
import com.nexosolar.android.databinding.FragmentFilterBinding
import com.nexosolar.android.domain.models.InvoiceFilters
import com.nexosolar.android.domain.models.InvoiceState
import com.nexosolar.android.ui.common.RangeValidator
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import java.time.Instant
import java.time.ZoneOffset

/**
 * Fragment para configurar filtros de facturas.
 * MIGRADO A FLOW: Usa StateFlow y repeatOnLifecycle.
 */
@AndroidEntryPoint
class FilterFragment : Fragment() {

    private var _binding: FragmentFilterBinding? = null
    private val binding get() = _binding!!

    // Referencia al ViewModel compartido
    private val viewModel: InvoiceViewModel by activityViewModels()

    /**
     * Estado temporal de filtros editados por el usuario.
     * Solo afecta al ViewModel al presionar "Aplicar".
     */
    private var tempFilters: InvoiceFilters? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentFilterBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupObservers()
        setupListeners()
    }

    // ========== OBSERVERS (MIGRADO A FLOW) ==========

    private fun setupObservers() {
        // ✅ CAMBIO CLAVE: Usar lifecycleScope + repeatOnLifecycle
        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                // Observamos el estado de filtros
                viewModel.filterState.collect { uiState ->
                    // Guardamos referencia inicial si es null
                    if (tempFilters == null) {
                        tempFilters = uiState.filters
                    }
                    renderUI(uiState)
                }
            }
        }
    }

    // ========== LISTENERS ==========

    private fun setupListeners() {
        with(binding) {
            btnSelectDate.setOnClickListener { showDatePicker(isStartDate = true) }
            btnSelectDateUntil.setOnClickListener { showDatePicker(isStartDate = false) }

            rangeSlider.addOnChangeListener { slider, _, _ ->
                handleSliderChange(slider.values)
            }

            setupStateCheckboxes()

            btnAplicar.setOnClickListener { applyFilters() }
            btnBorrar.setOnClickListener {
                // 1. Obtenemos el valor máximo real del ViewModel para reiniciar el slider correctamente
                val currentStats = viewModel.filterState.value.statistics

                // 2. Creamos unos filtros "limpios" (sin fechas, sin estados, rango completo)
                val cleanFilters = InvoiceFilters(
                    minAmount = 0f,
                    maxAmount = currentStats.maxAmount, // Importante: volver al tope real
                    startDate = null,
                    endDate = null,
                    filteredStates = emptySet()
                )

                // 3. Actualizamos la variable temporal del fragmento
                tempFilters = cleanFilters

                // 4. Forzamos el repintado de la pantalla con estos datos limpios
                // Truco: Creamos un estado "ficticio" temporal solo para que renderUI sepa qué pintar
                val dummyState = viewModel.filterState.value.copy(filters = cleanFilters)
                renderUI(dummyState)
            }
            btnCerrar.setOnClickListener { closeFragment() }
        }
    }

    private fun setupStateCheckboxes() {
        val checkboxMapping = mapOf(
            binding.checkPagadas to InvoiceState.PAID,
            binding.checkPendientesPago to InvoiceState.PENDING,
            binding.checkAnuladas to InvoiceState.CANCELLED,
            binding.checkCuotaFija to InvoiceState.FIXED_FEE,
            binding.checkPlanPago to InvoiceState.PAYMENT_PLAN
        )

        checkboxMapping.forEach { (checkbox, state) ->
            checkbox.setOnCheckedChangeListener { _, isChecked ->
                handleStateToggle(state, isChecked)
            }
        }
    }

    // ========== UI RENDERING ==========

    private fun renderUI(uiState: InvoiceFilterUIState) {
        // Usamos tempFilters si existe (edición en curso), sino el del estado
        val filtersToRender = tempFilters ?: uiState.filters

        with(binding) {
            renderDateButtons(filtersToRender)

            if (uiState.statistics.maxAmount > 0) {
                renderSlider(uiState, filtersToRender)
            }

            renderStateCheckboxes(filtersToRender)

            btnAplicar.isEnabled = !uiState.isApplying
        }
    }

    private fun renderDateButtons(filters: InvoiceFilters) {
        binding.btnSelectDate.text = filters.startDate?.let {
            DateUtils.formatDateShort(it)
        } ?: getString(R.string.dia_mes_ano)

        binding.btnSelectDateUntil.text = filters.endDate?.let {
            DateUtils.formatDateShort(it)
        } ?: getString(R.string.dia_mes_ano)
    }

    private fun renderSlider(uiState: InvoiceFilterUIState, currentFilters: InvoiceFilters) {
        val maxAmount = uiState.statistics.maxAmount

        with(binding.rangeSlider) {
            valueFrom = 0f
            valueTo = maxAmount

            val minVal = (currentFilters.minAmount ?: 0f).coerceIn(0f, maxAmount)
            val maxVal = (currentFilters.maxAmount ?: maxAmount).coerceIn(minVal, maxAmount)

            // Evitar crash si min > max por redondeo
            if (minVal <= maxVal) {
                setValues(minVal, maxVal)
            }
        }

        // Actualizar etiquetas
        binding.tvMaxImporte.text = String.format("%.2f €", maxAmount)
        binding.tvMinValue.text = formatSliderValue(currentFilters.minAmount ?: 0f, maxAmount, isMaxValue = false)
        binding.tvMaxValue.text = formatSliderValue(currentFilters.maxAmount ?: maxAmount, maxAmount, isMaxValue = true)
    }

    private fun renderStateCheckboxes(filters: InvoiceFilters) {
        val states = filters.filteredStates

        // Evitamos disparar listeners al setear estado programáticamente
        with(binding) {
            checkPagadas.isChecked = states.contains(InvoiceState.PAID.serverValue)
            checkPendientesPago.isChecked = states.contains(InvoiceState.PENDING.serverValue)
            checkAnuladas.isChecked = states.contains(InvoiceState.CANCELLED.serverValue)
            checkCuotaFija.isChecked = states.contains(InvoiceState.FIXED_FEE.serverValue)
            checkPlanPago.isChecked = states.contains(InvoiceState.PAYMENT_PLAN.serverValue)
        }
    }

    // ========== EVENT HANDLERS ==========

    private fun handleSliderChange(values: List<Float>) {
        if (values.size != 2) return

        val min = values[0]
        val max = values[1]

        tempFilters = tempFilters?.copy(minAmount = min, maxAmount = max) ?: InvoiceFilters(minAmount = min, maxAmount = max)

        // Feedback visual inmediato
        val maxAmount = viewModel.filterState.value.statistics.maxAmount
        binding.tvMinValue.text = formatSliderValue(min, maxAmount, isMaxValue = false)
        binding.tvMaxValue.text = formatSliderValue(max, maxAmount, isMaxValue = true)
    }

    private fun handleStateToggle(state: InvoiceState, isChecked: Boolean) {
        val currentStates = tempFilters?.filteredStates?.toMutableSet() ?: mutableSetOf()

        if (isChecked) {
            currentStates.add(state.serverValue)
        } else {
            currentStates.remove(state.serverValue)
        }

        // Crear filtros vacíos si es null
        val baseFilters = tempFilters ?: InvoiceFilters()
        tempFilters = baseFilters.copy(filteredStates = currentStates)
    }

    private fun showDatePicker(isStartDate: Boolean) {
        val uiState = viewModel.filterState.value
        // Usamos tempFilters si existe, si no los del estado
        val currentFilters = tempFilters ?: uiState.filters

        // --- 1. Definir Límites (Constraints) ---
        val minDateAllowed = uiState.statistics.oldestDateMillis
        val maxDateAllowed = uiState.statistics.newestDateMillis

        // (Opcional) Ajustar límites según la OTRA fecha ya seleccionada
        // Ej: Si elijo Fin, el mínimo no es oldestDate, es startDate (si existe)
        val validationMin = if (!isStartDate && currentFilters.startDate != null) {
            DateUtils.toEpochMilli(currentFilters.startDate).coerceAtLeast(minDateAllowed)
        } else minDateAllowed

        val validationMax = if (isStartDate && currentFilters.endDate != null) {
            DateUtils.toEpochMilli(currentFilters.endDate).coerceAtMost(maxDateAllowed)
        } else maxDateAllowed


        // --- 2. Definir Selección Inicial (Segura) ---
        val rawSelection = if (isStartDate) {
            currentFilters.startDate?.let { DateUtils.toEpochMilli(it) }
        } else {
            currentFilters.endDate?.let { DateUtils.toEpochMilli(it) }
        }

        // CLAVE: Si la fecha seleccionada cae FUERA del rango permitido, hacemos fallback
        // Esto evita el crash y muestra el calendario en una fecha válida
        val safeSelection = if (rawSelection != null && rawSelection in minDateAllowed..maxDateAllowed) {
            rawSelection
        } else {
            // Fallback inteligente:
            // Si es Fecha Inicio -> Pon el foco en el día más antiguo disponible
            // Si es Fecha Fin -> Pon el foco en el día más reciente disponible
            if (isStartDate) minDateAllowed else maxDateAllowed
        }

        // --- 3. Construir Picker ---
        val constraints = CalendarConstraints.Builder()
            .setStart(minDateAllowed)
            .setEnd(maxDateAllowed)
            .setValidator(RangeValidator(validationMin, validationMax)) // Tu validador custom
            .build()

        val picker = MaterialDatePicker.Builder.datePicker()
            .setTitleText(if (isStartDate) "Seleccionar Inicio" else "Seleccionar Fin")
            .setCalendarConstraints(constraints)
            .setSelection(safeSelection) // ¡Usamos la fecha segura!
            .build()

        picker.addOnPositiveButtonClickListener { selectedMillis ->
            handleDateSelected(selectedMillis, isStartDate)
        }

        picker.show(parentFragmentManager, "DATE_PICKER")
    }


    private fun handleDateSelected(dateMillis: Long, isStartDate: Boolean) {
        val selectedDate = Instant.ofEpochMilli(dateMillis)
            .atZone(ZoneOffset.UTC)
            .toLocalDate()

        val baseFilters = tempFilters ?: InvoiceFilters()
        tempFilters = if (isStartDate) {
            baseFilters.copy(startDate = selectedDate)
        } else {
            baseFilters.copy(endDate = selectedDate)
        }

        // Forzar renderizado parcial de botones
        tempFilters?.let { renderDateButtons(it) }
    }

    private fun applyFilters() {
        tempFilters?.let { filters ->
            viewModel.updateFilters(filters)
            viewModel.applyFilters()
        }
        closeFragment()
    }

    private fun closeFragment() {
        parentFragmentManager.popBackStack()
    }

    // ========== HELPERS ==========

    private fun formatSliderValue(value: Float, maxValue: Float, isMaxValue: Boolean): String {
        return if (isMaxValue && value >= maxValue - 0.001f) {
            String.format("%.2f €", value)
        } else {
            String.format("%.0f €", value)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}
